src/main/java/producer/ProducerRecord.java

```diff
@@ -0,0 +1,156 @@
+        this.partitionNumber = Optional.ofNullable(partition);
+        this.key = Optional.ofNullable(key);
+    }
+
+    public ProducerRecord(String topic, Integer partition, Long timestamp, K key, V value, Iterable<Header> headers) {
+        this.topic = topic;
+        this.timestamp = timestamp;
+        this.value = value;
+        this.partitionNumber = Optional.ofNullable(partition);
+        this.key = Optional.ofNullable(key);
+
+        if (headers != null) {
+            List<Header> headerList = new ArrayList<>();
+            headers.forEach(headerList::add);
+            this.headers = List.copyOf(headerList);
+        } else {
+            this.headers = Collections.emptyList();
+        }
+
+    }
+
+    public ProducerRecord(String topic, Integer partition, K key, V value) {
+        this.topic = topic;
+        this.value = value;
+        this.timestamp = LocalDateTime.now()
+                .atZone(ZoneId.systemDefault())
+                .toInstant()
+                .toEpochMilli();
+        this.partitionNumber = Optional.ofNullable(partition);
+        this.key = Optional.ofNullable(key);
+    }
+
+    public ProducerRecord(String topic, Integer partition, K key, V value, Iterable<Header> headers) {
+        this.topic = topic;
+        this.value = value;
+        this.timestamp = LocalDateTime.now()
+                .atZone(ZoneId.systemDefault())
+                .toInstant()
+                .toEpochMilli();
+        this.partitionNumber = Optional.ofNullable(partition);
+        this.key = Optional.ofNullable(key);
+        if (headers != null) {
+            List<Header> headerList = new ArrayList<>();
+            headers.forEach(headerList::add);
+           this.headers = List.copyOf(headerList);
+        } else {
+            this.headers = Collections.emptyList();
+        }
+    }
+
+    public ProducerRecord(String topic, K key, V value) {
+        this.topic = topic;
+        this.value = value;
+        this.timestamp = LocalDateTime.now()
+                .atZone(ZoneId.systemDefault())
+                .toInstant()
+                .toEpochMilli();
+        this.key = Optional.ofNullable(key);
+        if (this.key.isPresent()) {
+            this.partitionNumber = Optional.of(Math.abs(key.hashCode()));
+        }
+    }
+
+    public ProducerRecord(String topic, V value) {
+        this.topic = topic;
+        this.value = value;
+        this.timestamp = LocalDateTime.now()
+                .atZone(ZoneId.systemDefault())
+                .toInstant()
+                .toEpochMilli();
+        // TODO: Round Robin Implementation for partition number below;
+        this.partitionNumber = Optional.empty();
+    }
+
+    // METHODS
+    public String getTopic() {
+        return topic;
+    }
+
+    //TODO: Add Headers class.
+
+    public K getKey() {
+        return key.orElse(null);
+    }
+
+    public Integer getPartitionNumber() {
+        return partitionNumber.orElse(null);
```

Comment:
Wondering if you can just pass back the partition number instead of null. Esenstially just:

```
Optional<Integer> getPartitionNumber()
```

Reply:
Are you saying that because `partitionNumber` is always set with a value in every constructor?

Reply:
Mainly because of the `orElse(null)`, might as well just return the field itself. It would be different if it was something like `.orElse(90)` or `.orElse(10001)`